Sources/BSON/Integers+BSONValue.swift

```diff
@@ -1,6 +1,44 @@
 import NIO
 
 extension Int32: BSONValue {
+    /// Initializes an `Int32` given well-formatted canonical or relaxed extended JSON representing an `Int32`
+    /// Returns `nil` if the provided value is not an Int32.
+    /// Throws if the JSON is a partial match or is malformed.
+    internal init?(fromExtJSON json: JSON) throws {
+        switch json {
+        case let .number(n):
+            // relaxed extended JSON
+            guard let int = Int32(exactly: n) else {
+                return nil
+            }
+            self = int
+        case let .object(obj):
+            // canonical extended JSON
+            guard let value = obj["$numberInt"]?.stringValue else {
+                return nil
+            }
+            guard obj.count == 1 else {
+                throw DecodingError.dataCorrupted(
+                    DecodingError.Context(
+                        codingPath: [],
+                        debugDescription: "Not a valid Int32"
+                    )
+                )
+            }
+            guard let int = Int32(value) else {
+                throw DecodingError.dataCorrupted(
+                    DecodingError.Context(
+                        codingPath: [],
+                        debugDescription: "Not a valid Int32"
```

Comment:
Similarly, this could be a little more specific: "Could not parse a valid Int32 from \\"\\(value)\\""

Reply:
done